_posts/2020-10-19-torch-image-classification/torch-image-classification.Rmd

```diff
@@ -0,0 +1,543 @@
+
+-   How to use learning rate schedulers, and in particular, the one-cycle learning rate algorithm [\@abs-1708-07120].
+
+-   How to find a good initial learning rate.
+
+For convenience, the code is available on [Google Colaboratory](https://colab.research.google.com/drive/1OJzzqiQVbh3ZdLB2L2t_DhBGInlh9o-k?usp=sharing) -- no copy-pasting required.
+
+## Data loading and preprocessing
+
+The example dataset used here is available on [Kaggle](https://www.kaggle.com/gpiosenka/100-bird-species/data){.uri}.
+
+Conveniently, it may be obtained using [torchdatasets](https://github.com/mlverse/torchdatasets), which uses [`pins`](https://github.com/rstudio/pins) for authentication, retrieval and storage. To enable `pins` to manage your Kaggle downloads, please follow the instructions [here](https://pins.rstudio.com/articles/boards-kaggle.html).
+
+This dataset is very "clean", unlike the images we may be used to from, e.g., [ImageNet](http://image-net.org/). To help with generalization, we introduce noise during training -- in other words, we perform *data augmentation*. In `torchvision`, data augmentation is part of an *image processing pipeline* that first converts an image to a tensor, and then applies any transformations such as resizing, cropping, normalization, or various forms of distorsion.
+
+Here are the transformations to be performed on the training set. Note how most of them are for data augmentation, while normalization is done to comply with what's expected by ResNet.
+
+#### Image preprocessing pipeline
+
+```{r}
+library(torch)
+library(torchvision)
+library(torchdatasets)
+
+library(dplyr)
+library(pins)
+library(ggplot2)
+
+device <- if (cuda_is_available()) torch_device("cuda:0") else "cpu"
+
+train_transforms <- function(img) {
+  img %>%
+    # first convert image to tensor
+    transform_to_tensor() %>%
+    # then move to the GPU (if available)
+    (function(x) x$to(device = device)) %>%
+    # data augmentation
+    transform_random_resized_crop(size = c(224, 224)) %>%
+    # data augmentation
+    transform_color_jitter() %>%
+    # data augmentation
+    transform_random_horizontal_flip() %>%
+    # normalize according to what is expected by resnet
+    transform_normalize(mean = c(0.485, 0.456, 0.406), std = c(0.229, 0.224, 0.225))
+}
+
+```
+
+On the validation set, we don't want to introduce noise, while still needing to resize, crop, and normalize the images. The test set should be treated identically.
+
+```{r}
+valid_transforms <- function(img) {
+  img %>%
+    transform_to_tensor() %>%
+    (function(x) x$to(device = device)) %>%
+    transform_resize(256) %>%
+    transform_center_crop(224) %>%
+    transform_normalize(mean = c(0.485, 0.456, 0.406), std = c(0.229, 0.224, 0.225))
+}
+
+test_transforms <- valid_transforms
+```
+
+And now, let's get the data. A one-liner each will download the three sets and tell the corresponding R objects what transformations they're expected to apply: [^1]
```

Comment:
Does a one-liner download all three datasets or do you need a one-liner for each set? The order of the words here is throwing me a bit.

Reply:
for each ... i know it sounds weird, will try to find sth else still...